Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp

```diff
@@ -444,36 +435,42 @@ namespace Terrain
         }
 
         // macro clipmap data:
-        bool hasUpdate = false;
         for (uint32_t clipmapIndex = 0; clipmapIndex < m_macroClipmapStackSize; ++clipmapIndex)
         {
-            ClipmapBoundsRegionList updateRegionList = UpdateCenter(m_macroClipmapBounds[clipmapIndex], m_clipmapData.m_macroClipmapCenters[clipmapIndex]);
+            ClipmapBounds& clipmapBounds = m_macroClipmapBounds[clipmapIndex];
+
+            ClipmapBoundsRegionList updateRegionList = clipmapBounds.UpdateCenter(currentViewPosition);
 
-            for (uint32_t i = 0; i < ClipmapBounds::MaxUpdateRegions; ++i)
+            // write updated center
+            Vector2i center = clipmapBounds.GetModCenter();
+            m_clipmapData.m_clipmapCenters[clipmapIndex][0] = aznumeric_cast<uint32_t>(center.m_x);
+            m_clipmapData.m_clipmapCenters[clipmapIndex][1] = aznumeric_cast<uint32_t>(center.m_y);
+
+            AZ::Vector2 centerWorld = clipmapBounds.GetCenterInWorldSpace();
+            m_clipmapData.m_clipmapWorldCenters[clipmapIndex][0] = centerWorld.GetX();
+            m_clipmapData.m_clipmapWorldCenters[clipmapIndex][1] = centerWorld.GetY();
+
+            for (uint32_t i = 0; i < updateRegionList.size(); ++i)
             {
-                if (updateRegionList.size() > i)
-                {
-                    AZStd::array<uint32_t, 4> aabb = { (uint32_t)updateRegionList[i].m_localAabb.m_min.m_x,
-                                                       (uint32_t)updateRegionList[i].m_localAabb.m_min.m_y,
-                                                       (uint32_t)updateRegionList[i].m_localAabb.m_max.m_x,
-                                                       (uint32_t)updateRegionList[i].m_localAabb.m_max.m_y };
-                    m_clipmapData.m_macroClipmapBoundsRegions[clipmapIndex + ClipmapConfiguration::MacroClipmapStackSizeMax * i] = aabb;
-                }
-                else
-                {
-                    m_clipmapData.m_macroClipmapBoundsRegions[clipmapIndex + ClipmapConfiguration::MacroClipmapStackSizeMax * i] = zero;
-                }
+                AZStd::array<uint32_t, 4> aabb = { (uint32_t)updateRegionList[i].m_localAabb.m_min.m_x,
+                                                   (uint32_t)updateRegionList[i].m_localAabb.m_min.m_y,
+                                                   (uint32_t)updateRegionList[i].m_localAabb.m_max.m_x,
+                                                   (uint32_t)updateRegionList[i].m_localAabb.m_max.m_y };
```

Comment:
Nit: should use aznumeric_cast<uint32_t> instead of c-style cast.

Also, as a side note, is it correct to cast here? should the max use ceil()? are there any problems with it always truncating?

Reply:
It's a int to uint cast. The logic is already from the update, and it should always be >= 0. I think the original update uses int because we only have Vector2i (which is surprisingly defined in the Terrain Gem, we definitely need some better support for the basic types.)